src/bindings.jl

```diff
@@ -344,7 +344,11 @@ function add_binding(x, state, scope=state.scope)
         elseif scopehasbinding(scope, name)
             # TODO: some checks about rebinding of consts
             check_const_decl(name, b, scope)
-            scope.names[name] = b
+
```

Comment:
Not sure this is quite right e.g. 
```julia
function ff(arg1::T)
    arg1 = 2
    arg1
end
```
This leaves the final `arg1` pointing to the function argument rather than `= 2`.

Reply:
Did you mean 
```
if is_soft_scope(scope) && scopehasbinding(scope.parent, name)
    add_binding(scope.parent, binding) 
else
    scope.names[name] = b
end
```
? Because otherwise nothing changes about your example.

And yes, that's kinda annoying. But the issue with overwriting the binding is that `resolve_ref` then doesn't have the required info to find "references" before the binding was overwritten, which it should be able to do in my mind (so that renaming `arg1` in your example always renames all three instances).
Maybe the solution to that would be to add a kwarg to `resolve_ref` that optionally takes another pass over the scope expression to collect all "references". So basically we'd have the notion of "strict references" (everything after the binding was introduced) and "loose references" (everything in the same scope or a child that shares a name).

Reply:
I think I probably did mean that. Your suggestion sounds right - we need the interal representation to handle strict refs but it wouldn't be costly to do a pass to pick up all loose refs in response to an e.g. rename request